docs/man/mysqlfailover.1
'\" t
.\"     Title: \fBmysqlfailover\fR
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 01/14/2017
.\"    Manual: MySQL Utilities
.\"    Source: MySQL 1.6.5
.\"  Language: English
.\"
.TH "\FBMYSQLFAILOVER\FR" "1" "01/14/2017" "MySQL 1\&.6\&.5" "MySQL Utilities"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
mysqlfailover \- Automatic replication master failover
.SH "SYNOPSIS"
.HP \w'\fBmysqlfailover\ 'u
\fBmysqlfailover [\fR\fB\fIoptions\fR\fR\fB]\fR
.SH "DESCRIPTION"
.PP
This utility permits users to perform replication health monitoring and automatic failover on a replication topology consisting of a single master and its slaves\&. The utility is designed to run interactively or continuously refreshing the health information and checking the master status at periodic intervals\&. Its primary mission is to monitor the master for failure and when a failure occurs, execute failover to one of the slaves that is in a valid state\&. The utility accepts an optional list of slaves to be considered for the candidate slave\&.
.PP
This utility is designed to work exclusively for servers that support global transaction identifiers (GTIDs) and have
gtid_mode=ON\&. MySQL server versions 5\&.6\&.5 and higher support GTIDs\&. See
\m[blue]\fBReplication with Global Transaction Identifiers\fR\m[]\&\s-2\u[1]\d\s+2
for more information\&. Thus, this utility does not work with anonymous replication servers (binary log + position)\&.
.PP
The user can specify the interval in seconds to use for detecting the master status and generating the health report using the
\fB\-\-interval\fR
option\&. At each interval, the utility checks to see if the server is alive via a ping operation followed by a check of the connector to detect if the server is still reachable\&. The ping operation can be controlled with the
\fB\-\-ping\fR
option (see below)\&.
.PP
If the master is found to be offline or unreachable, the utility executes one of the following actions based on the
\fB\-\-failover\-mode\fR
option value\&. The available values are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBauto\fR
(default): Execute automatic failover to the list of candidates first and if no slaves are viable, continue to search the remaining slaves for a viable candidate\&. The command tests each candidate slave listed for the prerequisites\&. Once a candidate slave is elected, it is made a slave of each of the other slaves thereby collecting any transactions executed on other slaves but not the candidate\&. In this way, the candidate becomes the most up\-to\-date slave\&. If no slave is found to be a viable candidate, the utility generates an error and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBelect\fR: This mode is the same as auto, except if no candidates specified in the list of candidate slaves are viable, then it does not check the remaining slaves, and instead generates an error and then exits\&. Use this option to force failover to one or more specific slaves using the
\fB\-\-candidates\fR
option\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBfail\fR: This mode produces an error and does not failover when the master is detected as down or unreachable\&. This mode is used to provide periodic health monitoring without the failover action taken\&.
.RE
.PP
For all options that permit specifying multiple servers, the options require a comma\-separated list of connection parameters in the following form (where the password, port, and socket are optional)\&.:
.sp
.if n \{\
.RS 4
.\}
.nf
\fI*user*\fR[:\fI*passwd*\fR]@\fI*host*\fR[:\fI*port*\fR][:\fI*socket*\fR] or
\fI*login\-path*\fR[:\fI*port*\fR][:\fI*socket*\fR]
.fi
.if n \{\
.RE
.\}
.PP
The utility permits users to discover slaves connected to the master\&. The discover slaves feature is run automatically on each interval\&. Furthermore, it is required that slaves use the \-\-master\-info\-repository=TABLE startup setting\&.
.PP
The discover slaves option
\fBrequires\fR
all slaves use the
\-\-report\-host
and
\-\-report\-port
server startup variables with the correct hostname and port\&. If these are missing or report the incorrect information, the slave may not be detected and thus not included in the operation of the utility\&. The discover slaves option ignores any slaves to which it cannot connect\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
If you have one or more slaves which do not report their hostname and port and should a failover event occur, those slaves are not included in the resulting topology\&. That is, they are not a slave of the new master\&. Be sure to check that all of your slaves are accounted for in the health report before relying on the utility for complete automatic failover\&.
.sp .5v
.RE
.PP
The utility permits the user to specify an external script to execute before and after the switchover and failover commands\&. The user can specify these with the
\fB\-\-exec\-before\fR
and
\fB\-\-exec\-after\fR
options\&. The return code of the script is used to determine success\&. Each script must report 0 (success) to be considered successful\&. If a script returns a value other than 0, the result code is presented in an error message\&.
.PP
The utility also permits the user to specify a script to be used for detecting a downed master or an application\-level event to trigger failover\&. This can be specified using the
\fB\-\-exec\-fail\-check\fR
option\&. The return code for the script is used to invoke failover\&. A return code of 0 indicates failover should not take place\&. A return code other than 0 indicates failover should take place\&. This is checked at the start of each interval if a script is supplied\&. The timeout option is not used in this case and the script is run once at the start of each interval\&.
.PP
The utility permits the user to log all actions taken during the commands\&. The
\fB\-\-log\fR
option requires a valid path and filename of the file to use for logging operations\&. The log is active only when this option is specified\&. The option
\fB\-\-log\-age\fR
specifies the age in days that log entries are kept\&. The default is seven (7) days\&. Older entries are automatically deleted from the log file (but only if the
\fB\-\-log\fR
option is specified)\&.
.PP
The format of the log file includes the date and time of the event, the level of the event (informational \- INFO, warning \- WARN, error \- ERROR, critical failure \- CRITICAL), and the message reported by the utility\&.
.PP
The interface provides a number of options for displaying additional information\&. You can choose to view the replication health report (default), or choose to view the list of GTIDs in use, the UUIDs in use, or view the log file contents if logging is enabled\&. Each of these reports is described below\&.
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBhealth\fR
Display the replication health of the topology\&. This report is the default view for the interface\&. By default, this includes the host name, port, role (MASTER or SLAVE) of the server, state of the server (UP = is connected, WARN = not connected but can ping, DOWN = not connected and cannot ping), the GTID_MODE, and health state\&.
.sp
The master health state is based on the following: if GTID_MODE=ON, the server must have the binary log enabled, and a user must exist with the REPLICATE SLAVE privilege\&.
.sp
The
\fB\-\-seconds\-behind\fR
option is used to detect when a slave is behind the master\&. It allows users to set a threshold for reporting purposes only\&. It does not apply to slave candidacy or selection during failover\&.
.sp
The slave health state is based on the following: the IO_THREAD and SQL_THREADS must be running, it must be connected to the master, there are no errors, the slave delay for non\-GTID enabled scenarios is not more than the threshold provided by the
\fB\-\-max\-position\fR
and the slave is reading the correct master log file, and slave delay is not more than the
\fB\-\-seconds\-behind\fR
threshold option\&.
.sp
At each interval, if the discover slaves option was specified at startup and new slaves are discovered, the health report is refreshed\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBgtid\fR: Display the master\*(Aqs list of executed GTIDs, contents of the GTID variables;
@@GLOBAL\&.GTID_EXECUTED,
@@GLOBAL\&.GTID_PURGED, and
@@GLOBAL\&.GTID_OWNED\&. Thus, you can toggle through the four screens by pressing the \*(AqG\*(Aq key\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBUUID\fR: Display universally unique identifiers (UUIDs) for all servers\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBLog\fR: This option displays the contents of the log file, which only visible if the
\fB\-\-log\fR
option is specified\&. This can be helpful to see when failover occurred, and which actions or messages were recorded at the time\&.
.RE
.PP
The user interface is designed to match the size of the terminal window in which it is run\&. A refresh option is provided to permit users to resize their terminal windows or refresh the display at any time\&. However, the interface automatically resizes to the terminal window on each interval\&.
.PP
The interface displays the name of the utility, the master\*(Aqs status including binary log file, position, and filters as well as the date and time of the next interval event\&.
.PP
The interface also permits the user to scroll up or down through a list longer than what the terminal window permits\&. When a long list is presented, the scroll options become enabled\&. The user can scroll the list up with the up arrow key and down with the down arrow key\&.
.PP
Use the
\fB\-\-verbose\fR
option to see additional information in the health report and additional messages during failover\&.
.RE
.SS "MODES OF OPERATION"
.PP
The utility supports two modes of operation\&. The default mode, running as a console, works as described above\&. An additional mode that permits you to run the utility as a daemon is provided for POSIX platforms\&.
.PP
When run as a daemon, the utility does not have interactivity\&. However, all events are written to the log file\&. You can control what is written to the log by using the
\fB\-\-report\-values\fR
option\&.
.PP
To run the utility as a daemon, use the
\fB\-\-daemon\fR
option\&. There are four commands that can be used in
\fB\-\-daemon\fR
option\&. These include:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
start
.sp
Starts the daemon\&. The
\fB\-\-log\fR
option is required\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
stop
.sp
Stops the daemon\&. If you used the option
\fB\-\-pidfile\fR, the value must be the same when starting the daemon\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
restart
.sp
Restarts the daemon\&. If you used the option
\fB\-\-pidfile\fR, the value must be the same when starting the daemon\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
nodetach
.sp
Starts the daemon, but does not detach the process from the console\&. The
\fB\-\-log\fR
option is required\&.
.RE
.SS "OPTIONS"
.PP
\fBmysqlfailover\fR
accepts the following command\-line options:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-\-help
.sp
Display a help message and exit\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-\-license
.sp

CHANGES.txt
Release-1.6.6-IDA
-------------
- COMT#c8dca61 : Add INFORMATION_SCHEMA-based event schema object handling

- COMT#befe5bd : ISSUE-36: Avoid using SHOW DATABASES for alive check
- COMT#7d4f198 : no mysql.event in mysql 8, remove it
- COMT#4e7bd0a : no more mysql.proc in mysql 8, use information_schema.routines
- COMT#43ac066 : use the pure python connector implementation
- COMT#34f707e : Include traceback output when error occurs
- COMT#20e6c0b : Properly quote the definer for CREATE / ALTER VIEW statements
- COMT#7a1c8b6 : sql_transform.py: fix skip-opt-autoinc
- COMT#513139a : mysqldiff: make the result sql more readable
- COMT#40e0b99 : mysqldiff: add --include-create & --include-drop options
- COMT#e74f19d : mysqldiff: add --skip-opt-autoinc
- COMT#65843c1 : sql_transform.py:_get_index_sql_clauses(): make indexes by name and index fields in consistent order


Release-1.6.5
-------------
- BUG#25030151 : mysqlfailover (v:1.6) was stopped silently when using a '--exec-before' script
- BUG#24742356 : typos in mysqlfailover messages
- BUG#24657793 : mysqldbcopy fails to copy views which refer other views
- BUG#24624716 : ISSUE WITH 'MYSQLREPLICATE' 1.6.4 UTILITY
- BUG#24613794 : MYSQLFRM UNABLE TO PARSE SOME .FRM FILES
- BUG#23598948 : tables with primary key of 0 with auto_increment fail to copy
- BUG#23578714 : fix the _conn_ipv4_num_only regular expression
- BUG#23335753 : mysqldbcopy issue with tinytext
- BUG#22738248 : mysqlfailover fails with master my.cnf[entry] style option and python 2.6
- BUG#22616847 : incorrect error message from mysqlrpladmin
- BUG#22543157 : mysqlbinlogpurge: some documentation requests
- BUG#22543116 : mysqlbinlogpurge requires the master port explicitly
- BUG#22457829 : mysql serverclone doesn't detect the error properly and fails with ambiguous msg
- BUG#22254445 : --verbosity on mysqfrm does't work
- BUG#22215626 : mysqlserverinfo requires connection option if show-servers used
- BUG#22155217 : mysqldbimport --skip does not work
- BUG#22093518 : mysqldbcopy fails to copy grants for test database
- BUG#22093453 : mysqldbcompare --quiet does not work
- BUG#21848362 : mysqlfailover ping option description
- BUG#21764003 : convenience notation not permitted
- BUG#21626041 : ls: cannot access /usr/share/pyshared/mysql: no such file or directory
- BUG#21497923 : mu1.5.5: message to be updated for invalid parameters to mysqlrplcheck.py
- BUG#21270853 : mysqlrplsync does not allow for a --exclude option
- BUG#21201454 : mysqlfailover do not set read_only to slave during preparing candidate promotion
- BUG#21068133 : add ssl connection section to clearly document ssl behavior
- BUG#20857151 : mysqldiff drops and recreates pk and indexes when it is not needed
- BUG#20807296 : remove use password() function (deprecated as of mysql 5.7.6)
- BUG#20524886 : mysqlfailover doesn't clean up failover_console table (ie:drop) in daemon mode
- BUG#20511130 : --force doesn't work for missing objects with mysqldiff
- BUG#20411237 : mysqlfrmreader must support new year data type
- BUG#20400663 : mysqldbcopy documented required permissions list is incomplete
- BUG#18763591 : using grid format header difference should be ignored for mysqldbcompare
- BUG#18755438 : pep8 issue on utilities code base
- BUG#18733683 : mysqldbexport table names in the output are not consistent
- BUG#18293794 : fix some errors on utilities code base and improve it

Release-1.6.4
-------------
- BUG#24348569 : utilities fail to properly detect 5.7.13 servers 
- BUG#23300260 : mysql utilities fail with "too many values to unpack" in _update_alias
- BUG#23111808 : mysqldbbcopy copy blob data when some field's value is null
- BUG#22932375 : ha utilties must set connection_timeout in c/py connection to short value
- BUG#22905132 : handling regexp for few of the utilities
- BUG#22878052 : add master outage delay for mysqlfailover
- BUG#22780854 : valueerror while using mysqlrplsync discover-slave option with host:port
- BUG#22764612 : valueerror: too many values to unpack " running mysqlrpladmin
- BUG#22619159 : special char in password causes error with mysqlrplshow
- BUG#22543517 : mysqlbinlogpurge does not work with other than the first replication channel
- BUG#22495193 : mysqldbexport does not dump bit fields correctly
- BUG#22393353 : mysql-utilities-1.5.6-osx10.9.dmg won't install under osx 10.11.2 (el capitan)
- BUG#22242601 : yum install of mysql-utilities-1.5.6-1.el6.src.rpm fails (documentation updated)
- BUG#22076859 : after pressing <delete> in mysql utility console the command buffer is corrupt
- BUG#21222808 : mysqldiff: encoding comments always make databases definitions look different
- BUG#17666275 : add authentication plugin info to mysqluserclone

Release-1.6.3
-------------
- BUG#22964855 : year needs updation in readme,license and mysqluc in mu 1.6.3 packages
- BUG#22781266 : exclude option in mysqldbexport does not seem to take complete effect
- BUG#22628187 : mysqldbcopy does not work with MySQL 5.7
- BUG#22617526 : mysqlfailover and mysqlrpladmin do not work with MySQL 5.7
- BUG#22608528 : mysqlrplsync does not work with MySQL 5.7
- BUG#21966831 : Typo in mysqlbinlogrotate documentation
- BUG#21905774 : missing man pages in rpm package 
- BUG#21845449 : Fix mysqldbcompare when comparing the same object on the same server
- BUG#21841353 : UnicodeEncodeError in mysqlauditgrep using unicode characters
- BUG#21839348 : AttributeError in mysqldbcompare using --all option
- BUG#21784538 : The --exclude option is not working for data in mysqldbexport
- BUG#21653496 : Daemon processes creates pidfiles with incorrect privileges
- BUG#21653352 : MySQL Utilities breaks when sql_mode="ANSI_QUOTES"
- BUG#21572065 : mysqldbcompare raise errors when the index definitions differs
- BUG#21343759 : Fix my_print_defaults to read .mylogin.cnf on MySQL 5.6.25
- BUG#20876820 : win msi amd64 installs in program files (x86) regardless is a 64 

scripts/mysqlfailover.py
#!/usr/bin/env python
#
# Copyright (c) 2012, 2016, Oracle and/or its affiliates. All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
#

"""
This file contains the replication slave administration utility. It is used to
perform replication operations on one or more slaves.
"""

import logging
import os
import signal
import sys
import traceback

from mysql.utilities.common.tools import check_python_version
from mysql.utilities import VERSION_FRM, VERSION_STRING
from mysql.utilities.exception import UtilError, UtilRplError
from mysql.utilities.command.rpl_admin import RplCommands, purge_log
from mysql.utilities.common.messages import (SCRIPT_THRESHOLD_WARNING,
                                             MSG_UTILITIES_VERSION)
from mysql.utilities.common.server import check_hostname_alias
from mysql.utilities.common.tools import check_connector_python
from mysql.utilities.common.topology import parse_topology_connections
from mysql.utilities.common.options import (add_failover_options,
                                            add_verbosity, add_rpl_user,
                                            add_ssl_options,
                                            check_server_lists,
                                            license_callback,
                                            UtilitiesParser,
                                            check_password_security,
                                            check_script_option)

# Check Python version compatibility
check_python_version()

# Constants
NAME = "MySQL Utilities - mysqlfailover "
DESCRIPTION = ("mysqlfailover - automatic replication health monitoring and "
               "failover")
USAGE = "%prog --master=root@localhost --discover-slaves-login=root " + \
        "--candidates=root@host123:3306,root@host456:3306 "
_DATE_FORMAT = '%Y-%m-%d %H:%M:%S %p'
_DATE_LEN = 22

# Check for connector/python
if not check_connector_python():
    sys.exit(1)

if __name__ == '__main__':
    # Setup a terminal signal handler for SIGNIT
    # Must use SetConsoleCtrlHandler function on Windows!
    # If posix, save old terminal settings so we can restore them on exit.
    try:
        # Only valid for *nix systems.
        # pylint: disable=C0413,C0411
        import termios
        old_terminal_settings = termios.tcgetattr(sys.stdin)
    except:
        # Ok to fail for Windows
        pass

    def set_signal_handler(func):
        """Set the signal handler.
        """
        # If posix, restore old terminal settings.
        if os.name == "nt":
            from ctypes import windll
            windll.kernel32.SetConsoleCtrlHandler(func, True)
        # Install SIGTERM signal handler
        else:
            signal.signal(signal.SIGTERM, func)

    # If ctypes present, we have Windows so define the exit with decorators
    try:
        # pylint: disable=C0413,C0411
        import ctypes

        @ctypes.WINFUNCTYPE(ctypes.c_int, ctypes.c_uint)
        def on_exit(signal, func=None):
            """Override the on_exit callback.
            """
            logging.info("Failover console stopped with SIGTERM.")
            sys.exit(0)
    except:
        def on_exit(signal, func=None):
            """Override the on_exit callback.
            """
            if os.name == "posix":
                termios.tcsetattr(sys.stdin, termios.TCSADRAIN,
                                  old_terminal_settings)
            logging.info("Failover console stopped with SIGTERM.")
            sys.exit(0)

    set_signal_handler(on_exit)

    # Setup the command parser
    program = os.path.basename(sys.argv[0]).replace(".py", "")
    parser = UtilitiesParser(
        version=VERSION_FRM.format(program=program),
        description=DESCRIPTION,
        usage=USAGE,
        add_help_option=False,
        prog=program
    )

    # Default option to provide help information
    parser.add_option("--help", action="help",
                      help="display this help message and exit")

    # Add --License option
    parser.add_option("--license", action='callback',
                      callback=license_callback,
                      help="display program's license and exit")

    # Setup utility-specific options:
    add_failover_options(parser)

    # Interval for continuous mode
    parser.add_option("--interval", "-i", action="store", dest="interval",
                      type="int", default="15",
                      help="interval in seconds for polling the master for "
                           "failure and reporting health. Default = 15 "
                           "seconds. Lowest value is 5 seconds.")

    # Add failover modes
    parser.add_option("--failover-mode", "-f", action="store",
                      dest="failover_mode", type="choice", default="auto",
                      choices=["auto", "elect", "fail"],
                      help="action to take when the master fails. 'auto' = "
                           "automatically fail to best slave, 'elect' = fail "
                           "to candidate list or if no candidate meets "
                           "criteria fail, 'fail' = take no action and stop "
                           "when master fails. Default = 'auto'.")

    # Add failover detection extension point
    parser.add_option("--exec-fail-check", action="store", dest="exec_fail",
                      type="string", default=None,
                      help="name of script to execute on each interval "
                           "to invoke failover")

    # Add force to override registry entry
    parser.add_option("--force", action="store_true", dest="force",
                      default=False,
                      help="override the registration check on master for "
                           "multiple instances of the console monitoring the "
                           "same master.")

    # Add refresh script external point
    parser.add_option("--exec-post-failover", action="store",
                      dest="exec_post_fail", type="string", default=None,
                      help="name of script to execute after failover is "
                           "complete and the utility has refreshed the "
                           "health report.")

    # Pedantic mode for failing if some inconsistencies are found
    parser.add_option("-p", "--pedantic", action="store_true", default=False,
                      dest="pedantic",
                      help="fail if some inconsistencies are found (e.g. "
                           "errant transactions on slaves).")

    # Add no keyboard input
    parser.add_option("--no-keyboard", action="store_true", default=False,
                      dest="no_keyboard",
                      help="start with no keyboard input support.")

    # Add option to run as daemon
    parser.add_option("--daemon", action="store", dest="daemon", default=None,
                      help="run on daemon mode. It can be start, stop, "
                           "restart or nodetach.",
                      type="choice",
                      choices=("start", "stop", "restart", "nodetach"))

    # Add pidfile for the daemon option
    parser.add_option("--pidfile", action="store", dest="pidfile",
                      type="string", default=None,
                      help="pidfile for running mysqlfailover as a daemon.")

    # Add report values for daemon mode
    parser.add_option("--report-values", action="store", dest="report_values",
                      type="string", default="health",
                      help="report values used in mysqlfailover running as a "
                           "daemon. It can be health, gtid or uuid. Multiple "
                           "values can be used separated by commas. The "
                           "default is health.")

    # Add connection timeout for C/Py connections
    #
    # Note: we do not set a default here as we want C/Py to host the
    # default. Utilities need not control the default, rather, we only
    # want to provide an option for the user to set it should she have
    # need to tune for a particular failover scenario. See BUG#22932375
    # for more details.
    parser.add_option("--connection-timeout", action="store", type="int",
                      dest="conn_timeout", default=None, help="set the "
                      "connection timeout for TCP and Unix socket "
                      "connections for all master, slaves, and candidate "
                      "slaves specified. Default is 10 as provided in the "
                      "Connector/Python module.")

    # Add master failover delay check
    parser.add_option("--master-fail-retry", action="store", dest="fail_retry",
                      type="int", default=None, help="time in seconds to wait "
                      "to determine master is down. The failover check will "
                      "be run again when the retry delay expires. Can be used "
                      "to introduce a longer period between when master is "
                      "detected as unavailable to declaring it down. This "
                      "option is not used with --exec-fail-check. ")

    # Add verbosity mode
    add_verbosity(parser, False)

    # Replication user and password
    add_rpl_user(parser)

    # Add ssl options
    add_ssl_options(parser)

    # Now we process the rest of the arguments.
    opt, args = parser.parse_args()

    # Check security settings
    check_password_security(opt, args)

    # Check slaves list
    if opt.daemon != "stop":
        check_server_lists(parser, opt.master, opt.slaves)

    # Check for errors
    if int(opt.interval) < 5:
        parser.error("The --interval option requires a value greater than or "
                     "equal to 5.")

    # The value for --timeout needs to be an integer > 0.
    try:
        if int(opt.timeout) <= 0:
            parser.error("The --timeout option requires a value greater "
                         "than 0.")
    except ValueError:
        parser.error("The --timeout option requires an integer value.")

    # if opt.master is None and opt.daemon and opt.daemon != "stop":
    if opt.master is None and opt.daemon != "stop":
        parser.error("You must specify a master to monitor.")

    if opt.slaves is None and opt.discover is None and opt.daemon != "stop":
        parser.error("You must supply a list of slaves or the "
                     "--discover-slaves-login option.")

    if opt.failover_mode == 'elect' and opt.candidates is None:
        parser.error("Failover mode = 'elect' requires at least one "
                     "candidate.")

    if opt.fail_retry and opt.exec_fail:
        parser.error("The --master-fail-retry option cannot be used "
                     "with --exec-fail-check.")

    if opt.fail_retry and opt.fail_retry < 1:
        parser.error("The --master-fail-retry option must be a positive "
                     "integer > 0.")

    # Parse the master, slaves, and candidates connection parameters
    try:
        master_val, slaves_val, candidates_val = parse_topology_connections(
            opt)
    except UtilRplError:
        _, e, _ = sys.exc_info()
        print("ERROR: {0}".format(e.errmsg))
        sys.exit(1)

    # Check hostname alias
    for slave_val in slaves_val:
        if check_hostname_alias(master_val, slave_val):
            parser.error("The master and one of the slaves are the same "
                         "host and port.")
    for cand_val in candidates_val:
        if check_hostname_alias(master_val, cand_val):
            parser.error("The master and one of the candidates are the same "
                         "host and port.")

    # Create dictionary of options
    options = {
        'candidates': candidates_val,
        'ping': 3 if opt.ping is None else opt.ping,
        'verbosity': 0 if opt.verbosity is None else opt.verbosity,
        'before': opt.exec_before,
        'after': opt.exec_after,
        'exec_fail': opt.exec_fail,
        'max_position': opt.max_position,
        'max_delay': opt.max_delay,
        'discover': opt.discover,
        'timeout': int(opt.timeout),
        'interval': opt.interval,
        'failover_mode': opt.failover_mode,
        'logging': opt.log_file is not None,
        'log_file': opt.log_file,
        'force': opt.force,
        'post_fail': opt.exec_post_fail,
        'rpl_user': opt.rpl_user,
        'pedantic': opt.pedantic,
        'no_keyboard': opt.no_keyboard,
        'daemon': opt.daemon,
        'pidfile': opt.pidfile,
        'report_values': opt.report_values,
        'script_threshold': opt.script_threshold,
        'fail_retry': opt.fail_retry,
    }

    # Purge log file of old data
    if opt.log_file is not None and not purge_log(opt.log_file, opt.log_age):
        parser.error("Error purging log file.")

    # Setup log file
    logging.basicConfig(filename=opt.log_file, level=logging.INFO,
                        format='%(asctime)s %(levelname)s %(message)s',
                        datefmt=_DATE_FORMAT)

    # Warn user about script threshold checking.
    if opt.script_threshold:
        print(SCRIPT_THRESHOLD_WARNING)

    # Check if script files exist and are executable and warn users if they
    # are not.
    script_opts = ['after', 'before', 'exec_fail', 'post_fail']
    for key in script_opts:
        parameter_val = options[key]
        check_script_option(parser, parameter_val)

    # Check if the values specified for the --report-values option are valid.
    for report in opt.report_values.split(','):
        if report.lower() not in ("health", "gtid", "uuid"):
            parser.error("The value for the option --report-values is not "
                         "valid: '{0}', the values allowed are 'health', "
                         "'gitd' or 'uuid'".format(opt.report_values))

    # Check the daemon options
    if opt.daemon:
        # Check if a POSIX system
        if os.name != "posix":
            parser.error("Running mysqlfailover with --daemon is only "
                         "available for POSIX systems.")

        # Check the presence of --log
        if opt.daemon != "stop" and not opt.log_file:
            parser.error("The option --log is required when using --daemon.")

        # Test pidfile
        if opt.daemon != "nodetach":
            pidfile = opt.pidfile or "./failover_daemon.pid"
            pidfile = os.path.realpath(os.path.normpath(pidfile))
            if opt.daemon == "start":
                # Test if pidfile exists
                if os.path.exists(pidfile):
                    parser.error("pidfile {0} already exists. The daemon is "
                                 "already running?".format(pidfile))
                # Test if pidfile is writable
                try:
                    with open(pidfile, "w") as f:
                        f.write("{0}\n".format(0))
                    # Delete temporary pidfile
                    os.remove(pidfile)
                except IOError as err:
                    parser.error("Unable to write pidfile: {0}"
                                 "".format(err.strerror))
            else:
                # opt.daemon == stop/restart, test if pidfile is readable
                pid = None
                try:
                    if not os.path.exists(pidfile):
                        parser.error("pidfile {0} does not exist.".format(
                            pidfile))
                    with open(pidfile, "r") as f:
                        pid = int(f.read().strip())
                except IOError:
                    pid = None
                except ValueError:
                    pid = None
                # Test pid presence
                if not pid:
                    parser.error("Can not read pid from pidfile.")

    if opt.pidfile and not opt.daemon:
        parser.error("The option --daemon is required when using --pidfile.")

    # Log MySQL Utilities version string
    if opt.log_file:
        logging.info(MSG_UTILITIES_VERSION.format(utility=program,
                                                  version=VERSION_STRING))
    try:
        rpl_cmds = RplCommands(master_val, slaves_val, options)
        if opt.daemon:
            rpl_cmds.auto_failover_as_daemon()
        else:
            rpl_cmds.auto_failover(opt.interval)
    except UtilError:
        _, e, _ = sys.exc_info()
        # log the error in case it was an usual exception
        logging.log(logging.CRITICAL, e.errmsg.strip(' '))
        print("ERROR: %s" % e.errmsg)
        print(traceback.format_exc())
        sys.exit(1)
    except KeyboardInterrupt:
        sys.exit(0)

    sys.exit(0)
